src/shared/lib/observables/uploadProgressObservable.ts

```diff
@@ -1,95 +1,108 @@
+  delay(ms: number): Promise<void>;
   reset(): void;
 }
 
 class UploadProgressObservable
   extends CommonObservable
   implements IUploadProgressObservableSetters
 {
-  private _totalActivities: number = -1;
-
-  private _currentActivity: number = -1; // zero-based
-
-  private _currentActivityName: string = '';
-
-  private _totalFilesInActivity: number = -1;
-
-  private _currentFile: number = -1; // zero-based
-
-  private _currentSecondLevelStep: SecondLevelStep | null = null;
+  private uploadProgress: UploadProgress;
 
   constructor() {
     super();
+    this.uploadProgress = {} as UploadProgress;
+    this.reset();
+  }
+  set currentSecondLevelSteKey(value: SecondLevelStep | null) {
+    throw new Error('Method not implemented.');
   }
 
-  public set totalActivities(value: number) {
-    this._totalActivities = value;
+  public set totalActivities(value: number | null) {
+    this.uploadProgress.totalActivities = value;
   }
 
-  public set currentActivity(value: number) {
-    this._currentActivity = value;
+  public set currentActivity(value: number | null) {
+    this.uploadProgress.currentActivity = value;
   }
 
-  public set currentActivityName(value: string) {
-    this._currentActivityName = value;
+  public set currentActivityName(value: string | null) {
+    this.uploadProgress.currentActivityName = value;
   }
 
-  public set totalFilesInActivity(value: number) {
-    this._totalFilesInActivity = value;
+  public set totalFilesInActivity(value: number | null) {
+    this.uploadProgress.totalFilesInActivity = value;
     this.notify();
   }
 
-  public set currentFile(value: number) {
-    this._currentFile = value;
+  public set currentFile(value: number | null) {
+    this.uploadProgress.currentFile = value;
     this.notify();
   }
 
-  public set currentSecondLevelStep(value: SecondLevelStep | null) {
-    this._currentSecondLevelStep = value;
+  public set currentSecondLevelStepKey(value: SecondLevelStep | null) {
+    this.uploadProgress.currentSecondLevelStepKey = value;
     this.notify();
   }
 
   public get totalActivities() {
-    return this._totalActivities;
+    return this.uploadProgress.totalActivities;
   }
 
   public get currentActivity() {
-    return this._currentActivity;
+    return this.uploadProgress.currentActivity;
   }
 
   public get currentActivityName() {
-    return this._currentActivityName;
+    return this.uploadProgress.currentActivityName;
   }
 
   public get totalFilesInActivity() {
-    return this._totalFilesInActivity;
+    return this.uploadProgress.totalFilesInActivity;
   }
 
   public get currentFile() {
-    return this._currentFile;
+    return this.uploadProgress.currentFile;
+  }
+
+  public get currentSecondLevelStepKey() {
+    return this.uploadProgress.currentSecondLevelStepKey;
   }
 
-  public get currentSecondLevelStep() {
-    return this._currentSecondLevelStep;
+  public async delay(ms: number) {
+    await wait(ms);
   }
```

Comment:
Are you sure this should be a part of the Observable instance? It doesn't refer to the instance through `this` or change anything besides applying a side effect. 
Please consider moving this out into a separate function to separate concerns

Reply:
1. To highlight - "wait" is part of the progress observable. 
2. If to remove / exclude these wait's (delays) then we'd not get the observable working as expected - js thread would be busy on the AnswersUploadService code like encryption and we'd get jumps e.g. from 1 to 3.
3. This is much better than having "wait" in the AnswersUploadService.

Reply:
Moved wait into useUploadProgress